spmd/compiler/api.py

```diff
@@ -358,10 +392,34 @@ def forward(
         self, *args: Tuple[object], **kwargs: Dict[str, object]
     ) -> object:
         if self._compiled_m is None:
+            input_set = set(args)
+
+            def gather_inputs_for_compilation(inps):
+                compile_inps = [
+                    x
+                    if not isinstance(x, torch.Tensor) or x not in input_set
+                    else DTensor.from_local(x, self._schema.mesh, [Shard(0)])
```

Comment:
what this is actually doing is basically create a sharded DTensor then convert it to replicate, then to local in order to capture the correct shape?

Reply:
Yep, the SPMD API assumes that the inputs to forward function are implicitly sharded. So I "undo" that in the first iteration, so that we can capture a single-device graph that contains the "full / logical" shapes baked in , and then proceed to the parallelization afterwards.